Add support matrix for Cognitive Services (#2484)
microsoft/BotFramework-WebChat: SPEECH.md

> Sample code in this article is optimized for modern browsers. You may need to use a [transpiler](https://en.wikipedia.org/wiki/Source-to-source_compiler) (e.g. [Babel](https://babeljs.io/)) to target a broader range of browsers.

## Support matrix

<table>
  <thead>
    <tr>
      <td></td>
      <td></td>
      <td><center>Chrome/Edge/Firefox<br />on desktop</center></td>
      <td><center>Chrome<br />on Android</center></td>
      <td><center>Safari<br />on Mac OS and iOS</center></td>
      <td><center>Web View<br />on iOS</center></td>
      <td><center>Web View<br />on Android</center></td>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>STT</td>
      <td>Basic recognition</td>
      <td><center>✔ 4.6</center></td>
      <td><center>✔ 4.6</center></td>
      <td><center>✔ 4.6</center></td>

>       <td><center><nobr>❌ *1</nobr></center></td>

Reviewer: ```suggestion
      <td><center><nobr>❌ <sup>*1</sup></nobr></center></td>
```

preferably with an anchor

    </tr>
    <tr>
      <td>TTS</td>
      <td>Override using "speak" property</td>
      <td><center>✔</center></td>
      <td><center>✔</center></td>
      <td><center>✔</center></td>
      <td><center><nobr>❓ *2</nobr></center></td>
      <td><center>✔</center></td>
    </tr>
    <tr>
      <td>TTS</td>
      <td>Adaptive Cards using "speak" property</td>
      <td><center>✔</center></td>
      <td><center>✔</center></td>
      <td><center>✔</center></td>
      <td><center><nobr>❓ *2</nobr></center></td>
      <td><center>✔</center></td>
    </tr>
    <tr>
      <td>TTS</td>
      <td>Interrupt synthesis when clicking on microphone button</td>
      <td><center>✔ 4.6</center></td>
      <td><center>✔ 4.6</center></td>
      <td><center>✔ 4.6</center></td>
      <td><center><nobr>❓ *2</nobr></center></td>
      <td><center>✔ 4.6</center></td>
    </tr>
    <tr>
      <td>TTS</td>
      <td>Synthesize activity with multiple attachments</td>
      <td><center>✔</center></td>
      <td><center>✔</center></td>
      <td><center>✔</center></td>
      <td><center><nobr>❓ *2</nobr></center></td>
      <td><center>✔</center></td>
    </tr>
  </tbody>
</table>

Notes:

> 1. Web View on iOS is not a full browser. It does not have audio recording capabilities, which is required for Cognitive Services

Reviewer: add anchors

> Sample code in this article is optimized for modern browsers. You may need to use a [transpiler](https://en.wikipedia.org/wiki/Source-to-source_compiler) (e.g. [Babel](https://babeljs.io/)) to target a broader range of browsers.

## Support matrix

> <table>

Reviewer: I would have preferred this as a Markdown table. :/ Oh well.

  <thead>
    <tr>
      <th></th>
      <th></th>
      <th colspan="2">Chrome/Edge<br />and Firefox<br />on desktop</th>
      <th colspan="2">Chrome<br />on Android</th>
      <th colspan="2">Safari<br />on macOS/iOS</th>
      <th colspan="2"><a href="https://developer.android.com/reference/android/webkit/WebView">Web View<br />on Android</a></th>
      <th colspan="2"><a href="https://developer.apple.com/documentation/webkit/wkwebview">Web View<br />on iOS</a></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>STT</td>
      <td>Basic recognition</td>
      <th>✔</th><td>4.6</td>
      <th>✔</th><td>4.6</td>
      <th>✔</th><td>4.6</td>
      <th>✔</th><td>4.6</td>

>       <th>❌</th><td><a href="#notes"><sup>*1</sup></a></td>

Reviewer: Link to individual numbers? the anchors don't appear to work either, just refreshes the page. I don't know if this is an issue from being a PR

      <td>Adaptive Cards using "speak" property</td>
      <th>✔</th><td>4.5</td>
      <th>✔</th><td>4.5</td>
      <th>✔</th><td>4.5</td>
      <th>✔</th><td>4.5</td>
      <th>❓</th><td><a href="#notes"><sup>*2</sup></a></td>
    </tr>
    <tr>
      <td>TTS</td>
      <td>Interrupt synthesis when clicking on microphone button (<a href="https://github.com/microsoft/BotFramework-WebChat/issues/2428">Bug</a>) (<a href="https://github.com/microsoft/BotFramework-WebChat/pull/2429">PR</a>)</td>
      <th>✔</th><td>4.6</td>
      <th>✔</th><td>4.6</td>
      <th>✔</th><td>4.6</td>
      <th>✔</th><td>4.6</td>
      <th>❓</th><td><a href="#notes"><sup>*2</sup></a></td>
    </tr>
    <tr>
      <td>TTS</td>
      <td>Synthesize activity with multiple attachments</td>
      <th>✔</th><td>4.2</td>
      <th>✔</th><td>4.2</td>
      <th>✔</th><td>4.2</td>
      <th>✔</th><td>4.2</td>
      <th>❓</th><td><a href="#notes"><sup>*2</sup></a></td>
    </tr>
  </tbody>
</table>

### Notes

1. [Web View on iOS](https://developer.apple.com/documentation/webkit/wkwebview) is not a full browser. It does not have audio recording capabilities, which is required for Cognitive Services

> 2. As speech recognition is not working, speech synthesis is not tested

Reviewer: ```suggestion
2. As speech recognition is not working (see above), speech synthesis is not tested
```
?

> 3. On Safari, user gesture is always required for opening microphone. Web Chat cannot programmatically open microphone when we receive an expecting input hint

Reviewer: ```suggestion
3. On Safari, user gesture is always required for opening the microphone. Web Chat is unable to programmatically open the microphone when we receive an expecting input hint
```

   - https://github.com/WebAudio/web-audio-api/issues/790
   - We are planning to unblock by changing UX, https://github.com/microsoft/BotFramework-WebChat/issues/2211

> 4. Web Chat currently do not support selecting different device for audio recording

Reviewer: ```suggestion
4. Web Chat currently does not support selecting a different device for audio recording
```

-  Browser must [support Web Audio API][web audio api support]
   -  Available on all modern browsers on desktop and mobile platform

#### Special considerations for Safari on macOS and iOS

> Safari requires additional permission granted _implicitly_ by the user. The user needs to perform an interaction (click/tap/type) before any audio clips can be played during the browser session.

Reviewer: wouldn't this be explicitly?

-  `"acceptingInput"`: Web Chat will do nothing after the bot's message is spoken
-  `"ignoringInput"`: Web Chat will explicitly close the microphone

> > Note: `expectingInput` is currently not working on Safari on both macOS and iOS.

Reviewer: ```suggestion
> Note: `expectingInput` is currently not working on Safari for both macOS and iOS.
```
